File: November 2024 Leetcode Solution/November-28.py

from collections import deque
from typing import List

class Solution:
    def minimumObstacles(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        dist = [[float('inf')] * n for _ in range(m)]
        dist[0][0] = 0
        
        q = deque([(0, 0)])
        
        dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        
        while q:
            i, j = q.popleft()
            cur_dist = dist[i][j]
            
            if i == m - 1 and j == n - 1:
                return cur_dist
            
            for dx, dy in dirs:
                x, y = i + dx, j + dy
                if 0 <= x < m and 0 <= y < n:
                    new_dist = cur_dist + grid[x][y]
                    if new_dist < dist[x][y]:
                        dist[x][y] = new_dist
                        if grid[x][y] == 1:
                            q.append((x, y)) 
                        else:
                            q.appendleft((x, y))  
        
        return dist[m - 1][n - 1] if dist[m - 1][n - 1] != float('inf') else -1

File: November 2024 Leetcode Solution/November-28.java
class Solution {
    public int minimumObstacles(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        int[][] dist = new int[m][n];
        for (int[] row : dist) {
            Arrays.fill(row, Integer.MAX_VALUE);
        }
        dist[0][0] = 0;
        
        Deque<int[]> q = new ArrayDeque<>();
        q.offer(new int[]{0, 0});
        
        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        
        while (!q.isEmpty()) {
            int[] cell = q.poll();
            int i = cell[0], j = cell[1];
            int curDist = dist[i][j];
            
            if (i == m - 1 && j == n - 1) {
                return curDist;
            }
            
            for (int[] dir : dirs) {
                int x = i + dir[0], y = j + dir[1];
                if (x >= 0 && x < m && y >= 0 && y < n) {
                    int newDist = curDist + grid[x][y];
                    if (newDist < dist[x][y]) {
                        dist[x][y] = newDist;
                        if (grid[x][y] == 1) {
                            q.offer(new int[]{x, y});
                        } else {
                            q.offerFirst(new int[]{x, y});
                        }
                    }
                }
            }
        }
        
        return dist[m - 1][n - 1] != Integer.MAX_VALUE ? dist[m - 1][n - 1] : -1;
    }
}
